linera-views/src/views/bucket_queue_view.rs

```diff
@@ -0,0 +1,700 @@
+        let keys = Self::pre_load(&context)?;
+        let values = context.read_multi_values_bytes(keys).await?;
+        Self::post_load(context, &values)
+    }
+
+    fn rollback(&mut self) {
+        self.delete_storage_first = false;
+        self.cursor = Cursor::new(&self.stored_indices);
+        self.new_back_values.clear();
+    }
+
+    async fn has_pending_changes(&self) -> bool {
+        if self.delete_storage_first {
+            return true;
+        }
+        if !self.stored_indices.is_empty() {
+            let Some((i_block, position)) = self.cursor.position else {
+                return true;
+            };
+            if i_block != 0 || position != self.stored_indices.position {
+                return true;
+            }
+        }
+        !self.new_back_values.is_empty()
+    }
+
+    fn flush(&mut self, batch: &mut Batch) -> Result<bool, ViewError> {
+        let mut delete_view = false;
+        assert_eq!(self.stored_data.len(), self.stored_indices.len());
+        if self.delete_storage_first {
+            let key_prefix = self.context.base_key();
+            batch.delete_key_prefix(key_prefix);
+            delete_view = true;
+        }
+        if self.stored_count() == 0 {
+            let key_prefix = self.context.base_key();
+            batch.delete_key_prefix(key_prefix);
+            self.stored_indices = StoredIndices::default();
+            self.stored_data.clear();
+        } else if let Some((i_block, position)) = self.cursor.position {
+            for block in 0..i_block {
+                let index = self.stored_indices.indices[block].1;
+                let key = self.get_index_key(index)?;
+                batch.delete_key(key);
+            }
+            self.stored_indices.indices.drain(0..i_block);
+            self.stored_data.drain(0..i_block);
+            self.cursor = Cursor {
+                position: Some((0, position)),
+            };
+            // We need to ensure that the first index is in the front.
+            let first_index = self.stored_indices.indices[0].1;
+            if first_index != 0 {
+                let size = self.stored_indices.indices[0].0;
+                self.stored_indices.indices[0] = (size, 0);
+                let key = self.get_index_key(first_index)?;
+                batch.delete_key(key);
+                let key = self.get_index_key(0)?;
+                let data0 = self.stored_data.first().unwrap().as_ref().unwrap();
+                batch.put_key_value(key, &data0)?;
+            }
+        }
+        if !self.new_back_values.is_empty() {
+            delete_view = false;
+            let mut unused_index = match self.stored_indices.indices.last() {
+                Some((_, index)) => index + 1,
+                None => 0,
+            };
+            let new_back_values: VecDeque<T> = std::mem::take(&mut self.new_back_values);
+            let new_back_values: Vec<T> = new_back_values.into_iter().collect::<Vec<_>>();
```

Comment:
There's also `VecDeque::make_consecutive`, which doesn't reallocate.

Reply:
There is no function `make_consecutive`. There is a `make_contiguous` but I could not get it to work.
Note that this code does not clone.

Reply:
(Sorry, that's what I meant. No worries, maybe I'll give it a try later. Not important.)